# Embedded ML - Lab 2.1: TensorFlow

In this lab you will learn the basics of one of the most developed and widely used ML libraries: TensorFlow. It implements many of the most important ML models and algorithms and has optimized back-ends for efficient execution on CPUs, GPUs, TPUs and other devices.

In this lab you are given some helper functions but you are expected to write most of the code and be able to explain it at a high level of abstraction and also to modify any part of it. This lab is important because a significant part of the course will use TensorFlow.

### Learning outcomes


* Explain the basic concepts associated with TensorFlow
* Use the basic workflow of TensorFlow to build a simple ML model
* Implement simple dense networks with TensorFlow and Keras
* Use some of the input handling functions of TensorFlow
* Implement a simple CNN with TensorFlow and Keras

### TensorFlow workflow
As in general with ML, in TensorFlow you have to get or preprocess the model inputs, train the model, run inference and evaluate results.

Here you should use TensorFlow to build a dense 4-layer network to classify items in the FASHION MNIST dataset. Explore a few different hidden-layer sizes and report the accuracy achieved.

Finally, export the model to a file and write a separate code that is able to load that model and run inference again.

In [ ]:
import tensorflow as tf
from tensorflow import keras
import numpy as np
import matplotlib.pyplot as plt
import os

In [ ]:
# Pre-process input dataset
fashion_mnist = keras.datasets.fashion_mnist
(train_images, train_labels), (test_images, test_labels) = fashion_mnist.load_data()

# Create the model
def create_model(
    hidden_layer_one : int = 128,
    hidden_layer_two : int = 128,
    hidden_layer_three : int = 128,
    learnig_rate : float = 0.001
    ):
  model = keras.Sequential([
    keras.layers.Flatten(input_shape=(28,28)),
    keras.layers.Dense(hidden_layer_one, activation=tf.nn.relu),
    keras.layers.Dense(hidden_layer_two, activation=tf.nn.relu),
    keras.layers.Dense(hidden_layer_three, activation=tf.nn.relu),
    keras.layers.Dense(10, activation=tf.nn.softmax)
  ])

  # Compile the model
  model.compile(optimizer = tf.keras.optimizers.Adam(learning_rate=learnig_rate),
                loss = 'sparse_categorical_crossentropy',
                metrics=['accuracy'])

  return model

# Train the model
model = create_model()
model.fit(train_images, train_labels, epochs=5)

# Evaluate functional performance
test_loss, test_acc = model.evaluate(test_images, test_labels)
print("test lost: ", test_loss, "\ntest acc: ", test_acc)


4422102/4422102 [==============================] - 2s 0us/step
Epoch 1/5
1875/1875 [==============================] - 11s 3ms/step - loss: 1.2767 - accuracy: 0.7513
Epoch 2/5
1875/1875 [==============================] - 7s 4ms/step - loss: 0.5309 - accuracy: 0.8171
Epoch 3/5
1875/1875 [==============================] - 5s 3ms/step - loss: 0.4704 - accuracy: 0.8339
Epoch 4/5
1875/1875 [==============================] - 6s 3ms/step - loss: 0.4292 - accuracy: 0.8459
Epoch 5/5
313/313 [==============================] - 1s 2ms/step - loss: 0.4561 - accuracy: 0.8392
test lost:  0.45611512660980225 
test acc:  0.8392000198364258


In [ ]:
# Save the model to a file
model.save_weights("mi_primera_red")

In [ ]:
#create models
models = {}
models_performance = {}
arquitecture = [
    (128, 256, 64, 0.001),
    (64, 128, 256, 0.001),
    (256, 64, 128, 0.001),
    (128, 64, 32, 0.001),
    (64, 64, 32, 0.001),
    (32, 64, 32, 0.001),
    (256,256,256, 0.001)
    ]
for i in range(1, len(arquitecture)+1):
  print(f"#----------------------modelo_{i}--------------------#")
  models[f"modelo_{i}"] = create_model(
      arquitecture[i-1][0],
      arquitecture[i-1][1],
      arquitecture[i-1][2],
      arquitecture[i-1][3]
      )
  models[f"modelo_{i}"].summary()
  models[f"modelo_{i}"].fit(train_images, train_labels, epochs=5)
  test_loss, test_acc = models[f"modelo_{i}"].evaluate(test_images, test_labels)
  models_performance[f"modelo_{i}"] = (test_loss, test_acc)
  print("test lost: ", test_loss, "\ntest acc: ", test_acc)

#----------------------modelo_1--------------------#
Model: "sequential_1"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 flatten_1 (Flatten)         (None, 784)               0         
                                                                 
 dense_4 (Dense)             (None, 128)               100480    
                                                                 
 dense_5 (Dense)             (None, 256)               33024     
                                                                 
 dense_6 (Dense)             (None, 64)                16448     
                                                                 
 dense_7 (Dense)             (None, 10)                650       
                                                                 
Total params: 150602 (588.29 KB)
Trainable params: 150602 (588.29 KB)
Non-trainable params: 0 (0.00 Byte)
___________________________

In [ ]:
for key in models_performance:
  print(f"{key} loss: {models_performance[key][0]} ; acc: {models_performance[key][1]}")
  models[key].save_weights(f"{key}")


modelo_1 loss: 0.45906367897987366 ; acc: 0.845300018787384
modelo_2 loss: 0.569682776927948 ; acc: 0.7760999798774719
modelo_3 loss: 0.4234577715396881 ; acc: 0.8555999994277954
modelo_4 loss: 0.44674092531204224 ; acc: 0.84579998254776
modelo_5 loss: 0.4647466838359833 ; acc: 0.8363999724388123
modelo_6 loss: 0.5000714063644409 ; acc: 0.8184000253677368
modelo_7 loss: 0.49231085181236267 ; acc: 0.8299999833106995


In [ ]:
# Load the saved model
new_model = create_model()
# Pre-process test inputs
new_model.load_weights("mi_primera_red")
# Verify functional performance
test_loss, test_acc = new_model.evaluate(test_images, test_labels)
print("test lost: ", test_loss, "\ntest acc: ", test_acc)

313/313 [==============================] - 1s 2ms/step - loss: 0.4561 - accuracy: 0.8392
test lost:  0.45611512660980225 
test acc:  0.8392000198364258


### CNNs with TensorFlow
Convolutional Neural Networks add another type of processing layers to extract image features that allow the model to indentify patterns for a much better accuracy results in computer vision applications.

Implement a CNN model to classify the FASHION MNIST dataset and compare the accuracy results with the previous dense model. Also report a comparison of the model size measuring the saved model file size and through an analytical estimation.

In [ ]:
# Pre-process input dataset
fashion_mnist = keras.datasets.fashion_mnist
(train_images, train_labels), (test_images, test_labels) = fashion_mnist.load_data()
train_images = train_images.reshape(60000,28,28,1)
train_images = train_images / 255.0
test_images = test_images.reshape(10000,28,28,1)
test_images = test_images / 255.0
# Create the CNN model

model_conv = keras.Sequential([
    tf.keras.layers.Conv2D(128, (3, 3), activation='relu', input_shape=(28, 28, 1)),
    tf.keras.layers.MaxPooling2D(2, 2),
    tf.keras.layers.Conv2D(64, (3, 3), activation='relu'),
    tf.keras.layers.MaxPooling2D(2, 2),
    tf.keras.layers.Flatten(),
    tf.keras.layers.Dense(128, activation=tf.nn.relu),
    tf.keras.layers.Dense(10, activation=tf.nn.softmax)
])

# Compile the CNN model
model_conv.compile(optimizer = tf.keras.optimizers.Adam(learning_rate=0.0001),
                loss = 'sparse_categorical_crossentropy',
                metrics=['accuracy'])
model_conv.summary()
# Train the model
model_conv.fit(train_images, train_labels, epochs=10)
# Evaluate functional performance
test_loss, test_acc = model_conv.evaluate(test_images, test_labels)
print("test lost: ", test_loss, "\ntest acc: ", test_acc)
# Save the model to a file
model_conv.save_weights("modelo_conv")

Model: "sequential_9"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 conv2d (Conv2D)             (None, 26, 26, 128)       1280      
                                                                 
 max_pooling2d (MaxPooling2  (None, 13, 13, 128)       0         
 D)                                                              
                                                                 
 conv2d_1 (Conv2D)           (None, 11, 11, 64)        73792     
                                                                 
 max_pooling2d_1 (MaxPoolin  (None, 5, 5, 64)          0         
 g2D)                                                            
                                                                 
 flatten_9 (Flatten)         (None, 1600)              0         
                                                                 
 dense_36 (Dense)            (None, 128)              

In [ ]:
def calculate_operations(layer):
    if isinstance(layer, tf.keras.layers.Conv2D):
        num_filters = layer.filters
        kernel_size = layer.kernel_size[0] * layer.kernel_size[1]
        input_size = layer.input_shape[1] * layer.input_shape[2] * layer.input_shape[3]
        output_size = layer.output_shape[1] * layer.output_shape[2] * layer.output_shape[3]
        num_operations = input_size * output_size * kernel_size * num_filters * 2 + output_size * num_filters
    elif isinstance(layer, tf.keras.layers.Dense):
        input_size = layer.input_shape[1]
        output_size = layer.output_shape[1]
        num_operations = input_size * output_size * 2 + output_size
    else:
        num_operations = 0  # unsupported layer
    return num_operations

# Example usage:
total_operations_CNN = sum(calculate_operations(layer) for layer in model_conv.layers)
total_operations = sum(calculate_operations(layer) for layer in model.layers)
print("Total operations for the model with out:", total_operations)
print("Total operations for the CNN model:", total_operations_CNN)

Total operations for the model with out: 269194
Total operations for the CNN model: 349291600522


In [ ]:
#model size
def model_size(archivo1, archivo2):
    tamano_archivo1 = os.path.getsize(archivo1)
    print(f"{archivo1} el tamaño es {tamano_archivo1}")
    tamano_archivo2 = os.path.getsize(archivo2)
    print(f"{archivo2} el tamaño es {tamano_archivo2}")
    # Comparamos los tamaños
    if tamano_archivo1 > tamano_archivo2:
        print(f"{archivo1} es más grande que {archivo2}")
    elif tamano_archivo1 < tamano_archivo2:
        print(f"{archivo2} es más grande que {archivo1}")
    else:    tf.keras.layers.Conv2D(128, (3, 3), activation='relu', input_shape=(28, 28, 1)),
    tf.keras.layers.MaxPooling2D(2, 2),
        print(f"{archivo1} y {archivo2} tienen el mismo tamaño")

archivo1 = "mi_primera_red.data-00000-of-00001"
archivo2 = "modelo_conv.data-00000-of-00001"

model_size(archivo1, archivo2)

mi_primera_red.data-00000-of-00001 el tamaño es 1620777
modelo_conv.data-00000-of-00001 el tamaño es 3378787
modelo_conv.data-00000-of-00001 es más grande que mi_primera_red.data-00000-of-00001


### Transfer learning and fine tuning
When you want to build a model but do not have enough data or resources to train a network with the accuracy you need, it possible to use a model that has been pre-trained on a large dataset and fine tune it with the target (smaller)dataset to solve the target classification problem.

Here you should use TensorFlow and Keras to download a pre-trained vision model from TensorFlow Hub (e.g. MobileNet V2), add a softmax classification layer and train it with a small subset of the Fashion MNIST dataset.

Compare runtimes and Top-1 accuracy of the resulting model with the dense and convolutional models previously built.

In [ ]:
import numpy as np
import matplotlib.pylab as plt
import tensorflow as tf
import tensorflow_hub as hub
import tensorflow_datasets as tfds


def format_image(image, label):
    # Add an additional dimension for channels (grayscale images have 1 channel)
    image = tf.expand_dims(image, axis=-1)
    # Resize the image to (224, 224)
    image = tf.image.resize(image, (224, 224)) / 255.0
    image = tf.image.grayscale_to_rgb(
    image, name=None
)
    return image, label

fashion_mnist = keras.datasets.fashion_mnist
(train_images, train_labels), (test_images, test_labels) = fashion_mnist.load_data()

train_images =train_images[1500:2100,:,:]
train_labels =train_labels[1500:2100]
test_images =test_images[:1000,:,:]
test_labels =test_labels[:1000]

print(f"{train_images.shape} este el tamaño del la imagen")

# Since you are using fashion_mnist dataset, you can get the number of examples and classes directly
num_examples = len(train_images)
num_classes = len(np.unique(train_labels))
print(num_examples)
print(num_classes)

# Convert your data into a tf.data.Dataset
train_data = tf.data.Dataset.from_tensor_slices((train_images, train_labels))
test_data = tf.data.Dataset.from_tensor_slices((test_images, test_labels))

# split the data in training, validation, and test datasets
BATCH_SIZE = 32
train_batches = train_data.shuffle(num_examples // 4).map(format_image).batch(BATCH_SIZE).prefetch(1)
validation_batches = train_data.map(format_image).batch(BATCH_SIZE).prefetch(1)
test_batches = test_data.map(format_image).batch(1)

# display the shape of our data
for image_batch, label_batch in train_batches.take(1):
    pass
print(f"{image_batch.shape} imagen bath size")

# download the pre-trained model and create a Keras meta-layer

module_selection = ("mobilenet_v2", 224, 1280)
handle_base, pixels, FV_SIZE = module_selection
MODULE_HANDLE ="https://tfhub.dev/google/tf2-preview/{}/feature_vector/4".format(handle_base)
IMAGE_SIZE = (pixels, pixels)
print("Using {} with input size {} and output dimension {}".format(MODULE_HANDLE, IMAGE_SIZE, FV_SIZE))

feature_extractor = hub.KerasLayer(MODULE_HANDLE,
                                   input_shape=IMAGE_SIZE + (3,),
                                   output_shape=[FV_SIZE],
                                   trainable=False)

print("Building model with", MODULE_HANDLE)

# build a new model adding a softmax layer

model = tf.keras.Sequential([
        feature_extractor,
        tf.keras.layers.Dense(56, activation=tf.nn.relu),
        tf.keras.layers.Dense(num_classes, activation='softmax')
])

model.summary()

# compile and train the new model

model.compile(optimizer='adam',
                  loss='sparse_categorical_crossentropy',
                  metrics=['accuracy'])

EPOCHS = 5
model.fit(train_batches,
                 epochs=EPOCHS,
                 validation_data=validation_batches)

4422102/4422102 [==============================] - 1s 0us/step
(600, 28, 28) este el tamaño del la imagen
600
10
(32, 224, 224, 3) imagen bath size
Using https://tfhub.dev/google/tf2-preview/mobilenet_v2/feature_vector/4 with input size (224, 224) and output dimension 1280
Building model with https://tfhub.dev/google/tf2-preview/mobilenet_v2/feature_vector/4
Model: "sequential"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 keras_layer (KerasLayer)    (None, 1280)              2257984   
                                                                 
 dense (Dense)               (None, 56)                71736     
                                                                 
 dense_1 (Dense)             (None, 10)                570       
                                                                 
Total params: 2330290 (8.89 MB)
Trainable params: 72306 (282.45 KB)
Non-trainable params: 2

In [ ]:

test_loss, test_acc = model.evaluate(test_batches)
print("test lost: ", test_loss, "\ntest acc: ", test_acc)


1000/1000 [==============================] - 11s 10ms/step - loss: 0.5713 - accuracy: 0.7890
test lost:  0.5712833404541016 
test acc:  0.7889999747276306


In [ ]:
model.save_weights("modelo_transfer_tunning")

In [ ]:
model.summary()

Model: "sequential"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 keras_layer (KerasLayer)    (None, 1280)              2257984   
                                                                 
 dense (Dense)               (None, 56)                71736     
                                                                 
 dense_1 (Dense)             (None, 10)                570       
                                                                 
Total params: 2330290 (8.89 MB)
Trainable params: 72306 (282.45 KB)
Non-trainable params: 2257984 (8.61 MB)
_________________________________________________________________


# Conclusion

The total number of operations for the dense model is 269,194.
The total number of operations for the convolutional neural network (CNN) model is 349,291,600,522. This indicates significantly higher computational complexity compared to the dense model, which is expected due to the additional convolutional and pooling layers in the CNN architecture.
Additionally, you've observed that the size of the saved weights for the CNN model is larger than that of the dense model. Specifically, the size of modelo_conv.data-00000-of-00001 is 3,378,787 bytes, while mi_primera_red.data-00000-of-00001 is 1,620,777 bytes.
These observations highlight the trade-offs between model complexity, computational requirements, and storage space. While CNNs can achieve higher accuracy in certain tasks, they often require more computational resources and storage due to their deeper architectures and larger number of parameters.

In this practice, we embarked on a journey through TensorFlow, a foundational tool in machine learning, to construct and train neural network models for image classification tasks using the Fashion MNIST dataset. Beginning with dense neural networks, we experimented with varying architectures and observed their impact on model accuracy. Transitioning to convolutional neural networks (CNNs), we witnessed their superior performance, attributed to their ability to capture spatial features in images. Through calculations of computational complexity and comparison of storage sizes, we grasped the trade-offs inherent in model design, balancing between accuracy and resource efficiency. This practice not only equipped us with practical skills in TensorFlow but also instilled a deeper understanding of fundamental concepts crucial for navigating the landscape of machine learning applications.

Using pre-trained models is a very useful way to save time and computation, since not having to train the entire network allows testing different types of classifiers with more varieties in significantly less time.  However, the weight of the file “model_transfer_tunning.data-00000-of-00001” is 9,970,911 bytes, which is the heaviest so far, this is because the pre-trained network brings a lot of information that is not specific to this problem and has some more general convolutional layers, which are not all used. Pre-trained algorithms 'saves time but at the cost of more memory depending on the pre-trained network used.

When deploying models in resource-constrained environments or considering storage limitations, it's essential to strike a balance between model complexity and performance. Techniques such as model compression, quantization, and architecture optimization can help mitigate these challenges while maintaining acceptable accuracy levels.